[JSC] GreedyRegAlloc: implement intra-block live range splitting

dhecht opened the pull request:
#### 823ded11d0caf68359fe9ac511e7c06d8c637e4f
<pre>
[JSC] GreedyRegAlloc: implement intra-block live range splitting
<a href="https://bugs.webkit.org/show_bug.cgi?id=303868">https://bugs.webkit.org/show_bug.cgi?id=303868</a>
<a href="https://rdar.apple.com/166163403">rdar://166163403</a>

Reviewed by Marcus Plutowski.

To give the register allocator more options when under register pressure,
implement live range splitting within a basic block. Within a block,
if a Tmp cannot be allocated to a register, then the allocator
attempts to create new Tmps to carry the values within use-def clusters
inside each basic block that the Tmp is accessed. This way, we have
a pattern in a basic block such as:

use Tmp
use Tmp
def Tmp
use Tmp

(in this example, the first cluster has 2 uses of Tmp and the second
cluster has a def and use of Tmp.)

and Tmp spills, currently the allocator will (in the worst case, ignoring
other existing optimizations) rewrite this to:

Mov (spill), r0
use r0
Move (spill), r1
use r1
def r2
Move r2, (spill)
Move (spill), r3
use r3

With intra-block live range splitting, the allocator now has the option
to write this as:

Move (spill), r0
use r0
use r0
def r1
Move r1, (spill) # emitted only if Tmp is live out of the block
use r1

Note that this has some overlap with the fixObviousSpills pass, which
can do a similar transformation when rN==rM, etc. But doing this
at register allocation time can find additional opportunities when
the spill tmps did not happen to get the same register.

Also note that fixObviousSpills can do some additional transformations
such as rewriting constant materializations as Adds and replacing
spill uses with registers across block boundaries. So both are useful.

In the future, this form of live range splitting can be enhanced to
split across loops rather than just intra-block, so spilled
tmps can potentially live in a register for the duration of the loop.

Canonical link: <a href="https://commits.webkit.org/304261@main">https://commits.webkit.org/304261@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0d074d9efef8d99e43642ee639ccd3d7bba59955

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135041 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/7448 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/138/builds/46294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142549 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/155/builds/8073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/7295 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/103189 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137987 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/155/builds/8073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/138/builds/46294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/84042 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/155/builds/8073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/138/builds/46294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3145 "Built successfully") | | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/127061 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/155/builds/8073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/138/builds/46294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145247 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/133538 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/7126 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/46294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111564 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/7177 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/7295 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/111924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/138/builds/46294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/61068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20832 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/7174 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/46294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/166419 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/6946 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/70749 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/43489 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/7177 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/7052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

Achierius commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -603,16 +603,60 @@ struct TmpData {
+    void add(Point instPoint)
+    {
+        if (m_instPoints.isEmpty() || m_instPoints.last() != instPoint) {
+            ASSERT(m_instPoints.isEmpty() || m_instPoints.last() < instPoint);
```
Could this be hoisted above to `ASSERT(m_instPoints.isEmpty() || m_instPoints.last() <= instPoint);`? More likely to hit the assert that way

dhecht replied:
I could, but then it'd have to be:
```
 ASSERT(m_instPoints.isEmpty() || m_instPoints.last() <= instPoint);
```
since we actively filter equal points. And so this won't be more likely to assert, it'd be equivalent. And I think the current assert is clearer as to what invariant we are maintaining (that points are monotonically increasing).

dhecht commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -1752,11 +1808,157 @@ class GreedyAllocator {
+    template<Bank bank>
+    bool trySplitIntraBlock(Tmp tmp, TmpData& tmpData)
+    {
+        CompilerTimingScope timingScope("Air"_s, "GreedyRegAlloc::trySplitIntraBlock"_s);
```
Oops, I meant to remove this. I will do so before merging.

Achierius commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -1752,11 +1808,155 @@ class GreedyAllocator {
+        return true;
+    }
+
+    void ensureUseDefLists()
```
I assume this would behave badly with changes to the tmps used by instructions -- e.g. if someone did something like

1. ensureUseDefLists() -- inadvertently/indirectly/etc.
2. <modify tmps>
3. ensureUseDefLists() -- thinking that this would take into account the modified tmps
4. <modify tmps again>

This feels like it'd be worth an assert (though unsure how exactly) or at least a comment

dhecht replied:
Once a tmp is spilled it's true that this instruction may not literally reference the Tmp anymore (instead it references the spill slot). But once that happens, that Tmp is no longer interesting to register allocation (Spilled is a terminal state through the Tmp state machine).

Also, it's not really a problem since this doesn't hold the use def references directly; it just tells you what instructions had use/def the Tmp at some point in time. I'm not sure how to assert this though, but I'll leave a comment to explain this a bit.

Achierius commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -1752,11 +1808,155 @@ class GreedyAllocator {
+    // Returns the unprocessed portion of the interval (if interval spans multiple blocks).
+    // `cursor` can be used to perform a "sort-merge join" when the caller is making queries over a sorted set of intervals for the same tmp
+    template<typename Func>
+    Interval forEachUseDefWithin(Tmp tmp, Interval interval, size_t& cursor, const Func& func)
```
The function seems to make a lot of assumptions about this cursor, which I think works as it's used right now but feels fragile -- e.g. you need to use the same cursor for all intervals, they have to be non-overlapping, and I believe strictly increasing which is stricter than just 'sorted' since that could be satisfied by a strictly-decreasing set?

It'd be nice if this could be some sort of generator instead but I understand why you didn't do so.
Maybe a middle-ground would be to make a wrapper IntervalIterator that could handle asserts (e.g. for the sortedness, non-intersection) and encapsulate cursor handling.

dhecht replied:
A fundamental property of LiveRanges is that the intervals are sorted and non overlapping intervals. So I think this is covered by `LiveRange::validate()` and also the ASSERT a few lines down from here about `cursor`. I agree it's not strongly abstracted but I'd prefer to not complicate it since this is all super internal the the greedy register allocator. But let me know if that misses your point.

Achierius commented on Source/JavaScriptCore/b3/air/AirUseCounts.h:
```diff
@@ -120,9 +120,9 @@ class UseCounts {
         if constexpr (bank == GP)
-            return m_gpConstDefs.quickGet(absoluteIndex);
+            return m_gpConstDefs.get(absoluteIndex);
         else
-            return m_fpConstDefs.quickGet(absoluteIndex);
+            return m_fpConstDefs.get(absoluteIndex);
```
What changed to make this necessary? Or was it a bug beforehand?

dhecht replied:
It's now necessary because with this change we now generate new Tmps (after the UseCounts has been generated) that can spill. Before this change, all Tmps generated during register allocation were "unspillable" or not referenced directly by existing IR, e.g. the gap Tmps for splitting around clusters could spill but they are referenced only by the fix up code inserted by register allocation).

So with these new spoilable Tmps that are referenced by existing IR (due to the rewrite of the tmpPtrs), the spilling code may query this for Tmps that are out of bounds. We could either expand these sets/vectors to include these new tmps, but doing that seems like a waste of memory and more complexity than it's worth (especially since this is called only from the register allocator).

Achierius commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -1777,23 +1977,28 @@ class GreedyAllocator {
         // Batch the generation of spill/fill tmps so that we can limit traversals of the code while
         // not tracking each tmp's use/defs explicitly.
-        m_didSpill = true;
+        m_needsEmitSpillCode = true;
```
Should this be tracked per-bank, given that m_didSpill is?

dhecht replied:
`m_didSpill` is now used to keep track of whether the bank had any spills. `m_needsEmitSpillCode` is used to track state while register allocating, but we do the banks serially (first all GP registers than all FP registers), so we only need one bool for the `m_needsEmitSpillCode` state.

Achierius commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -1752,11 +1808,155 @@ class GreedyAllocator {
+        if (tmpData.splitMetadataIndex)
+            return false;
+
+        unsigned tmpIndex = AbsoluteTmpMapper<bank>::absoluteIndex(tmp);
+        if (bank == GP && m_useCounts.isConstDef<bank>(tmpIndex))
+            return false; // Constant will be rematerialized instead
```
Am I correct to say that this isn't necessary for correctness, just a performance heuristic?

dhecht replied:
Yes, basically. We could handle this but it would take some added complexity and it's not clear it would be a win, so I went with the simpler way. The reason it adds complexity is because we remove the original defs for these "const def tmps", and so we'd need to complicate that decision based on whether some uses were split. See the `doKillInst` related logic; we'd have to make that smart enough to know about cluster tmp uses.

Achierius commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -1752,11 +1808,155 @@ class GreedyAllocator {
+        if (bank == GP && m_useCounts.isConstDef<bank>(tmpIndex))
+            return false; // Constant will be rematerialized instead
+
+        BasicBlock* startBlock = findBlockContainingPoint(tmpData.liveRange.intervals().first().begin());
+        Point last = tmpData.liveRange.intervals().last().end() - 1;
+        if (last <= positionOfTail(startBlock))
+            return false; // Tmp's live range is already block local
```
Whereas this is necessary for correctness?

dhecht replied:
This is to avoid infinite recursion of the newly created cluster tmps. There might be some cases where a tmp is already intra-block only but happens to have independent lifetimes and so could in theory benefit, but I don't think that happens in practice.

Achierius replied:
Probably could use a comment to distinguish it from the previous perf-only case

dhecht replied:
Done

Achierius commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -1752,11 +1808,155 @@ class GreedyAllocator {
+                        metadata = &m_splitMetadata.last();
+                    }
+                    Point pre = pointAtOffset(cluster.begin(), PointOffsets::Pre);
+                    // If the Tmp is live into the cluster then cluster range is extended to model the fixup load from
+                    // the Tmp's spill slot before the first use. This also allows insertSplitIntraBlockFixupCode()
+                    // to quickly determine whether the fixup load is needed.
+                    if (interval.contains(pre))
```
I don't understand why this extension is necessary... wouldn't that liveness already be captured in the live range you used to construct the cluster?

dhecht replied:
It needs to be extended at least to include the Early point but really including the Pre point is most accurate in terms of modeling the lifetime. The cluster tmp needs to be loaded before the cluster in this case, so we need to ensure it conflicts with any early use or defs of this instruction. And since the load happens in the gap between existing instructions, modeling it as starting at Pre makes sense. And has that nice side effect that the comment states which is to make it easy to determine if this load should be inserted during the fix up phase.

Achierius commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -2036,37 +2258,95 @@ class GreedyAllocator {
+        }
+    }
+
+    void insertSplitIntraBlockFixupCode(SplitMetadata& metadata)
```
When entering this is it possible for originalTmp's stage to be anything but Stage::Spilled?

dhecht replied:
No, it's not possible because we explicitly do `spill()` on that `Tmp` when intra-block splitting succeeds. That's not strictly necessary but it was the simplest way since otherwise we'd have to poke holes in the original Tmp's live range to try to allocate it. Which is doable and similar to what we do for "split around clobbers", but not something I wanted to take on in the first implementation.

Achierius commented on Source/JavaScriptCore/b3/air/AirAllocateRegistersByGreedy.cpp:
```diff
@@ -1837,6 +2042,22 @@ class GreedyAllocator {
+                    for (auto& split : metadata.splits) {
+                        Tmp clusterTmp = split.tmp;
+                        TmpData& clusterData = m_map[clusterTmp];
+                        if (clusterData.stage == Stage::Spilled) {
```
For my own understanding, would this ever not be true other than when the cluster is of size 1?

dhecht replied:
It's true whenever we tried to do intra-block splitting but still weren't able to find registers for one or more of these cluster tmps. Basically, for those clusters, we're giving up and just using the spill slot for the original tmp in that cluster.

Achierius reviewed (approved):
r=me

dhecht committed:
[JSC] GreedyRegAlloc: implement intra-block live range splitting
https://bugs.webkit.org/show_bug.cgi?id=303868
rdar://166163403

Reviewed by Marcus Plutowski.

To give the register allocator more options when under register pressure,
implement live range splitting within a basic block. Within a block,
if a Tmp cannot be allocated to a register, then the allocator
attempts to create new Tmps to carry the values within use-def clusters
inside each basic block that the Tmp is accessed. This way, we have
a pattern in a basic block such as:

use Tmp
use Tmp
def Tmp
use Tmp

(in this example, the first cluster has 2 uses of Tmp and the second
cluster has a def and use of Tmp.)

and Tmp spills, currently the allocator will (in the worst case, ignoring
other existing optimizations) rewrite this to:

Mov (spill), r0
use r0
Move (spill), r1
use r1
def r2
Move r2, (spill)
Move (spill), r3
use r3

With intra-block live range splitting, the allocator now has the option
to write this as:

Move (spill), r0
use r0
use r0
def r1
Move r1, (spill) # emitted only if Tmp is live out of the block
use r1

Note that this has some overlap with the fixObviousSpills pass, which
can do a similar transformation when rN==rM, etc. But doing this
at register allocation time can find additional opportunities when
the spill tmps did not happen to get the same register.

Also note that fixObviousSpills can do some additional transformations
such as rewriting constant materializations as Adds and replacing
spill uses with registers across block boundaries. So both are useful.

In the future, this form of live range splitting can be enhanced to
split across loops rather than just intra-block, so spilled
tmps can potentially live in a register for the duration of the loop.

Canonical link: https://commits.webkit.org/304261@main
